### DISCUSSION TITLE
What is the best way to have two isolated machines communicate with each other?

### DISCUSSION THREAD

**Discussion by zacharyweidenbach** [statelyai/xstate#3345] [Q&A] [answered]
First and explanation of why I want this:

The pattern that many react/xstate examples use (and I like) is to have individual machines for each screen of an application. However, I would like the concern of managing api data (both fetching, caching, and invalidating cache) to be separate from the concern of the render logic of the screens themselves. In that way, the screen's machine won't handle the data, it will only send an event that it needs the data to another machine that will be responsible for fetching and caching the data (as well as expose it to the UI. Let's call this fetch/cache machine 'QueryMachine'. Since this cache could be used in other areas of the app, I think it makes sense to treat this machine as a thing that's "always on". Because of that, I like the pattern of making these various 'QueryMachines' for different endpoints as singletons, and interpreting their machines at runtime to expose their `send` methods out to anything that wants to consume them.

A machine that control the UI, let's call it 'UIMachine' will depend on these singleton QueryMachine services in order to trigger the event to GET_DATA, and then wait until RECEIVED_DATA returns from the QueryMachines, and adjust its own state accordingly. 

However, there is a missing piece. If a command request (PUT POST PATCH DELETE) occurs that invalidates the cache of related QueryMachines, it's unclear the best way to cascade that back to the UI machines.

Here is a diagram of the layering I'm trying to describe:
![Screen Shot 2022-05-26 at 10 48 27 AM](https://user-images.githubusercontent.com/14946412/170548527-5ac6665b-5212-4e51-a3fb-32af6e1c4157.png)

A more detailed writeup of the specific use-case I'm using to illustrate this need can be found here:
https://github.com/statelyai/xstate/issues/366#issuecomment-1136727850

But the short version is:

> react-navigation keeps screens mounted after navigating forward in the same navigation stack. If machines are organized at the screen level, and some mutation occurs in a later screen that needs to cascade to the previous screen (which is still mounted) to reflect the change, it's unclear the best way to do that.
> 
> In this contrived example, let's say there are 2 screens:
> 
> ItemListScreen
> ItemEditScreen
> Let's also say you can only get to ItemEditScreen through ItemListScreen
> The ItemListScreen fetches a list of items. The ItemEditScreen will edit a single item via PATCH request to /items/:id
> Maybe you can see the problem already. If the ItemEditScreen mutates the data and we want that reflected on the list on the previous screen (which is still mounted btw), we need to manage that somehow.

Xstate doesn't seem to provide a clear way for isolated running machines to communicate with each other via events. The concept of Actors only seems to apply to machines with Parent/Child relationships. But due to the ephemeral nature of the UIMachine pattern, where the machine that controls the UI has it's lifecycle tied to the mounting and unmounting of the react component itself, that would seem to prohibit any kind of parent/child relationship, as the global api cache machines are already running, and not something that can be invoked and then spun down once the screen is done talking to it (as there might be other machines higher in the component tree that also need to talk to the same api cache machine). 

One approach I've tried is to use the `someInterpretedMachineService.subscribe` from inside another machine to establish a link, but this feels highly anti-pattern-y. Example below:
```ts
import { createMachine, interpret, assign } from 'xstate';

export const parentMachine = createMachine({
  id: 'parentMachine',
  initial: 'start',
  states: {
    start: {
      on: {
        PARENT_NEXT: 'next',
      },
    },
    next: {
      on: {
        PARENT_END: 'end',
      },
    },
    end: {},
  },
  on: {
    RESET: 'start',
  },
});

export const parentService = interpret(parentMachine);

const sendParentToNext = () => parentService.send('PARENT_NEXT');
const subscribeToParentNextEvent = (cb: () => void) =>
  parentService.subscribe((state) => {
    if (state.matches('next')) {
      cb();
    }
  });

export const childMachine = createMachine(
  {
    id: 'childMachine',
    context: {
      subscription: { unsubscribe: () => {} },
    },
    initial: 'start',
    entry: 'subscribeToParentNextEvent',
    exit: 'unsubscribeFromParentNextEvent',
    states: {
      start: {
        on: {
          CHILD_NEXT: {
            target: 'next',
            actions: 'sendParentToNext',
          },
        },
      },
      next: {
        on: {
          CHILD_END: 'end',
        },
      },
      end: { type: 'final' as const },
    },
    on: {
      RESET: 'start',
    },
  },
  {
    actions: {
      subscribeToParentNextEvent: assign({
        subscription: () => {
          return subscribeToParentNextEvent(() => {
            childService.send('CHILD_END');
          });
        },
      }),
      unsubscribeFromParentNextEvent: (context) => {
        context.subscription.unsubscribe();
      },
      sendParentToNext: () => {
        sendParentToNext();
      },
    },
  },
);

export const childService = interpret(childMachine);
```

Making the QueryMachines as true parents to the ephemeral UI machines feels worse though. Reasoning about a deeply nested machine that manages all the api cache machines _and_ all the ephemeral UI Machines and relays events between the two feels way too complex.

I think what I want is something that allows me to declare a machine's parent service/s before it is interpreted and started.
`machine.withParents([service1, service2])`

But maybe that introduces indeterminism to the machines as they won't know when a new machine might spontaneously become a child and begin receiving and sending events.

**Comment by zacharyweidenbach** [comment]
Inside the react component that calls `useMachine` on the UIMachine, there could be a `useEffect` that listens for changes to the QueryMachine service with a `useSelector` in order to send an event to the UIMachine to reset its own state, but using react as the connective tissue between machines seems to go against the philosophy of xstate being framework agnostic, as it entangles concerns about how machines listen to each other with react. And ideally, react is only responsible for UI rendering concerns.

**Comment by zacharyweidenbach** [comment]
The more I read through the discussions, the more it seems like the answer is "just organize your machines into a giant parent/child tree so you can send and receive events in all your machines." But then that raises a lot of questions. If the entry point of all machines starts with a single "supermachine" at the very top that then spawns children, who spawn their own children, etc. what does that even look like from a single screen in the app wanting to subscribe to its corresponding UIMachine? Does it use react context to reach into the "supermachine" and then do some deep query into its state to find the machine for its own screen? It seems like you'd need to be very careful about abstracting this in a way that you don't get overwhelmed with the complexity of the entire state hierarchy.

It seems like the rather elegant pattern of `useMachine` at the level of the screen that is controlled by that machine goes out the window because of the aforementioned reaching deep into the state of the supermachine to find the specific state node you care to listen to. Also, if you change the structure of your super machine, that could have massive cascading breakages on deeply nested machines trying to talk to a distant parent via events. Would the supermachine pattern also make testing pieces in isolation a lot harder to reason about when you need to create a parent machine testing harness so the machine under test doesn't break?

**Comment by Andarist** [comment]
> Xstate doesn't seem to provide a clear way for isolated running machines to communicate with each other via events. 

`send` action can target just any other running actor - you can only have to pass references around to make it possible, but that's inevitable. 

```ts
const machine = createMachine({
  on: {
    FOO: {
      actions: send({ type: 'FETCH_SMTH' }, { to: ctx => ctx.queryRef })
    }
  }
})

useMachine(machine, {
  context: {
    queryRef // might come from props or from the module's scope or from somewhere else
  }
})
```

**Reply by zacharyweidenbach** [reply]
Okay, this answers half of the problem, here's a codesandbox illustrating the issue:
https://codesandbox.io/s/recursing-architecture-e3u2i7?file=/src/machine.ts

If the singleton QueryMachine is interpreted, and then spawned by the UIMachine that wants to send and receive messages to it, the UIMachine can only _send_ messages to the QueryMachine. If the interpreted QueryMachine service wants to send an event back, it can't do that with `sendParent` because it was instantiated as a singleton service prior to being spawned as an actor, so I guess it thinks it doesn't have any parents. Maybe I could send along a ref for the QueryMachine to hold on to until it responds, but that doesn't solve the problem of needing to push events that originated from elsewhere down to listeners, like in the case of some CommandMachine invalidating the cache of one or many QueryMachines. Because of that, it feels very much like the QueryMachine is actually the parent and the UIMachine is the child. But that doesn't seem correct either, because a UIMachine might need to interact with many QueryMachines if it needs data from multiple sources, and I don't think a machine can have many parents.

I suppose I could keep some kind of subscription machine in front of the QueryMachine that manages who to push out to when events originate from the QueryMachine, but I'd rather not do that.

**Reply by Andarist** [reply]
You can grab `origin` of an event, like this:
```ts
assign((_, __, { origin }) => ({
  requesterRef: origin
}))
```
and then use this reference as the `send`'s target.

**Reply by zacharyweidenbach** [reply]
ah, so if I understand correctly, the QueryMachine would need to manage its requester refs, if it had many different machines attempting to "request" it concurrently, as well as when requesters unsubscribe from it when their machines exit? So it sounds like you don't get the ability to send/receive events to/from another already-running machine "for free" in this scenario, but there is a workaround. I would probably want to move out the management of requesters as it's own machine that would sit in front of the actual QueryMachine so responsibilities don't get muddled.

**Reply by Andarist** [reply]
> ah, so if I understand correctly, the QueryMachine would need to manage its requester refs, if it had many different machines attempting to "request" it concurrently, as well as when requesters unsubscribe from it when their machines exit? So it sounds like you don't get the ability to send/receive events to/from another already-running machine "for free" in this scenario,

Would you get such an ability "for free" with any other solution? I mean - some layer has to always manage such "requesters", otherwise you wouldn't be able to respond appropriately.

**Reply by zacharyweidenbach** [reply]
what I mean is, sending and receiving events "comes for free" with parent/child machines if I understand correctly.
Slight amendment to the code you posted based on my own observation
```ts
assign((_, __, assignMeta) => ({
  requesterRef: assignMeta._event.origin
}))
```

Followup question: When an already-running interpreted machine is spawned by another machine, is that itself a detectable event by the spawned service, such that it could capture the origin of the machine that spawned it? Put another way, when I spawn the SubscriptionMachine that manages connections to the QueryMachine, can I avoid having to send it some `SUBSCRIBE` and `UNSUBSCRIBE` event on the `entry` and `exit` of the UIMachine, and instead look at some implicit event that occurs from spawning the SubscriptionMachine, and exiting when the UIMachine tears down? It would cut down on boilerplate code if all the UIMachine needed to do was spawn the QueryMachine's SubscriptionMachine to begin sending and receiving events from it.

**Reply by zacharyweidenbach** [reply]
Actually it looks like this entire approach is still not going to work. Because the connection between the UIMachine and the singleton always-running QueryMachine is made by "spawning" the QueryMachine as a ref in the context of the ephemeral UIMachine, which treats the QueryMachine interpreted services as a child of the UIMachine. So when the UIMachine stops, it stops the QueryMachine interpreted services as well, and I want that to _always be on_ so other UIMachines that might still be running can talk to it.

What I want is actually really simple and it seems xstate is overcomplicating it by an order of magnitude. This is a simple event stream subscription. The QueryMachine has an event stream that it emits events into. And one or many UIMachines can subscribe to the event stream. UIMachines can emit events into the stream and the QueryMachine can detect these and perform a task accordingly (get the data, either from the cache or from a fetch), and emit a response with the new data. Likewise, the QueryMachine can emit events, like cache invalidation, that were sourced from somewhere else other than the subscribed UIMachine/s in order to push updates down to the UIMachine/s to reset their local state. At this point, it looks like I need to use a vanilla event emitter library to bridge the two machines so that one doesn't try to act like the parent of the other, but rather, just as a kind of network layer to send and receive events across an event stream.

**Reply by zacharyweidenbach** [reply]
@Andarist 

The solution I came up with is to invoke a subscription machine that subscribes to the event stream of the global QueryMachine. The QueryMachine emits events into this event stream as side effect actions while it does its thing. From here, it's easy to see how if something like a PUT/POST/PATCH/DELETE request needed to invalidate the store of another QueryMachine, it could just grab the QueryMachine's exposed send method and send a RESET event to it. The QueryMachine would then emit a RESET event into its event stream and if any machine downstream is subscribed, it will detect the event and send it up to the parent machine to do with as needed.

So a given UIMachine will look like:
![Screen Shot 2022-06-02 at 10 02 25 PM](https://user-images.githubusercontent.com/14946412/171789582-a024c8ac-1da5-4145-b7b6-b9be9a3e7f5e.png)

By using parallel states, a UIMachine can invoke one or many subscriptions and decide what local state to send the UIMachine to in response.

Here's a link to a test illustrating the RESET behavior
https://github.com/zacharyweidenbach/react-native-test-sandbox/blob/main/NativeFSM/store/utils/queryMachineWithColdStorage/index.integration.spec.ts#L134

And a link to the subscription machine factory service:
https://github.com/zacharyweidenbach/react-native-test-sandbox/blob/main/NativeFSM/store/utils/subscriptionMachineFactory.ts

So far, this approach entails a lot of boilerplate. For every new QueryMachine, there needs to be a new event stream, and SubscriptionMachine to subscribe to that QueryMachine's event stream, not to mention any machine subscribing needs parallel state. It feels bulky, even for such a trivial example app.

Basically what I've done is reinvent `react-query` just for the sake of divorcing the core application logic, in this case the network layer, away from React altogether. But after having done this, I might concede that using `react-query` in a hybrid approach with xstate is going to be more practical for getting up and running faster. On the other hand, doing things this way gives a level of control that you'd lose with `react-query`. For instance, I have made query machines that only persist data in memory, and ones that will persist to some cold storage repository (asyncStorage on mobile) and modular enough that I could slot in a new cold storage repository pretty easily. 

Anyways, thank you for taking time to respond to my ramblings. Xstate is awesome.

**Comment by zacharyweidenbach** [comment, marked as answer]
At the moment, the best answer I have come up with so far is detailed in this response:

https://github.com/statelyai/xstate/discussions/3345#discussioncomment-2875852

However, if there are more elegant solutions to the problem, I am all ears.

**Reply by ChrisShank** [reply]
You can make a small event bus [example](https://github.com/ChrisShank/xstate-behaviors/blob/master/examples/event-bus/main.ts), [source](https://github.com/ChrisShank/xstate-behaviors/blob/master/src/event-bus/from-event-bus.ts). It doesn't matter if they are in the same actor tree or not.

**Reply by zacharyweidenbach** [reply]
@ChrisShank 
Thank you for this example! This leverages the built in `subscribe` of the running machine, similar to my code example in the original post. But if I understand this approach, it still would not work for my needs. You use a ParentMachine to "link" two machines together with a new instance of an event bus that both can access. But to do that, the ParentMachine invokes both machines, which effectively "starts" both of them at the same time. What I want is for an ephemeral machine to "link" up to an already-running machine to begin sending and receiving events from it. This way, the lifecycle of the ephemeral machine is not tied to the already-running machine in any way.

In my example, the already-running machine is the QueryMachine which stores cached api data, as well and handles fetching/refetching data. The ephemeral machine is the UIMachine. It's lifecycle is tied to that of the screen when the user navigates there. It should be able to start up, subscribe to the QueryMachine, and then, when the user navigates away and the screen unmounts, it will unsubscribe but the QueryMachine remains running.

This approach also means that a machine could only subscribe to one other machine at a time since either machine can only have one parent. If there is a UIMachine that needs to subscribe to multiple QueryMachines, this would be a problem.

Hopefully I didn't misunderstand the behavior of you solution.

**Reply by ChrisShank** [reply]
No need to take my example at face value, you can share an event bus between two machines without the need of a parent!

```ts

const eventBus = new EventBus('EV')

const pingMachine = createMachine<Context>({
  id: 'ping',
  invoke: {
    id: 'event bus',
    src: fromEventBus(() => eventBus,
  },
  on: {
    PONG: {
      actions: [log('PONG'), send({ type: 'PING' }, { to: 'event bus', delay: 1000 })],
    },
  },
});

const pongMachine = createMachine<Context>({
  id: 'pong',
  invoke: {
    id: 'event bus',
    src: fromEventBus(() => eventBus),
  },
  on: {
    PING: {
      actions: [log('PING'), send({ type: 'PONG' }, { to: 'event bus', delay: 1000 })],
    },
  },
});

```

**Reply by zacharyweidenbach** [reply]
@ChrisShank 

Ah I see now! Sorry, I misunderstood the coupling between the EventBus instance and the parent machine saving it to its context. This feels a lot lighter weight and flexible than what I have now. Thank you for the additional example! I'm definitely going to use this. Also, I learned another thing from your example, which is that the `invoke` can take an array of functions to invoke in serial. This means that I don't need parallel states anymore to manage subscriptions between machines. The machine can simply invoke one or many event bus subscriptions and then have a monolithic `states` object. That feels a lot more natural than what I came up with.

Edit:
I've also starred your repo as it looks to have a lot of interesting/useful helpers and information for better understanding xstate.
